### PR TITLE
rust-analyzer: 2021-03-22 -> 2021-04-05

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change
- Version bump.
- Merge `default.nix` and `generic.nix` since they are often changed together when adding and removing patches.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 118833` at 2bd42e36 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>rust-analyzer</li>
    <li>rust-analyzer-unwrapped</li>
    <li>vscode-extensions.matklad.rust-analyzer</li>
  </ul>
</details>

File: pkgs/development/tools/rust/rust-analyzer/generic.nix [hunk 1]
```diff
@@ -17,6 +17,24 @@ rustPlatform.buildRustPackage {
...
+    })
+    (fetchpatch { # Use more std::array::IntoIter
+      url = "https://github.com/rust-analyzer/rust-analyzer/commit/bc5c86543b1bf384555471415dee770888a1781d.patch";
+      sha256 = "sha256-IfPe8dx2UQO+zMZ8u3qzVqDeyfcMhq46ZKEMQvDCzXA=";
+    })
+    (fetchpatch { # Use arrayvec 0.6
+      url = "https://github.com/rust-analyzer/rust-analyzer/commit/9787bddac577a6aa24388fb91286474a7a8cf0bc.patch";
+      sha256 = "sha256-pypbzMNJ8ywUp4pA+UQ3dsqrg2tVf0bNt4CE22w7aVk=";
+    })
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/tools/rust/rust-analyzer/generic.nix:36]: Those patches are already included in 2021-04-05. Do I miss something?

Comment (oxalica) [reply pkgs/development/tools/rust/rust-analyzer/generic.nix:36]: > Patches required before rust 1.51.0 to revert commits updating arrayvec.

Currently we have 1.50.0 in nixpkgs master, but these changes made it requires 1.51.0 to compile (1.51.0 is still in staging). This blocks rust-analyzer to some weeks. Here we **revert** some commits to made it compile on 1.50.0.

Comment (oxalica) [reply pkgs/development/tools/rust/rust-analyzer/generic.nix]: They are now removed for rustc 1.51.0 in master.

**Pull request merged**